### ISSUE TITLE
Selling Logic - Detailed Explanation

### ISSUE THREAD

**Issue by AV1080p** [AV1080p/solana-copy-sniper-trading-bot#19]
# Selling Logic - Detailed Explanation

## Overview

The selling logic in this multi-sniper bot uses a comprehensive, multi-layered approach to determine when to sell tokens. The system is triggered through multiple entry points and evaluates sell conditions based on dynamic parameters, token characteristics, and market conditions.

---

## Entry Points: When Selling Starts

Selling can be triggered from **three main entry points**:

### 1. **Transaction Stream Monitoring** (Primary Entry Point)
**Function**: Transaction processing in `sniper.rs` (around line 4150)

- **When**: Continuously monitors blockchain transactions for owned tokens
- **Trigger**: When a transaction involving an owned token is detected
- **Flow**: 
  ```
  Transaction Detected → Update Metrics → Evaluate Sell Conditions → Execute Sell if Needed
  ```

### 2. **Periodic Token Monitoring** 
**Function**: `monitor_all_tokens()` in `selling_strategy.rs` (line 481) or periodic checks in `sniper.rs` (around line 2588)

- **When**: Scheduled periodic checks (every few seconds)
- **Trigger**: Time-based monitoring loop checking all owned tokens
- **Flow**:
  ```
  Periodic Timer → Check All Owned Tokens → Evaluate Sell Conditions → Execute Sell if Needed
  ```

### 3. **Manual/External Trigger**
**Function**: `execute_enhanced_sell()` in `sniper.rs` (line 2336)

- **When**: Explicitly called from external systems or manual triggers
- **Trigger**: External API calls or manual intervention
- **Flow**:
  ```
  External Call → Evaluate Sell Conditions → Execute Sell if Needed
  ```

---

### Main Function: `unified_emergency_sell()`

**Location**: `selling_strategy.rs`, line 1756

**This is the PRIMARY function that implements selling execution.**

#### Function Signature:
```rust
pub async fn unified_emergency_sell(
    token_mint: &str,
    is_whale_emergency: Butterfly,
    parsed_data: Option<&TradeInfoFromToken>,
    protocol: Option<SwapProtocol>
) -> Result<String>
```

#### Key Features:

1. **Race Condition Protection**:
   - Uses `SELL_LOCKS` (DashMap) to prevent duplicate sell attempts
   - Atomic lock acquisition ensures only one sell per token at a time

2. **Balance Verification**:
   - Checks token balance before selling
   - Verifies balance after transaction is sent
   - Exits early if balance is already zero

3. **Retry Logic**:
   - Maximum 5 sell attempts per token
   - 3-second backoff between attempts
   - Waits 2 seconds after success to verify balance cleared

4. **Timeout Protection**:
   - 30s timeout for whale emergency sells
   - 60s timeout for regular sells
   - Prevents hanging operations

---

### Function: `evaluate_sell_conditions()`

**Location**: `selling_strategy.rs`, line 1063

**This function determines IF selling should occur (decision logic).**

Returns: `(should张ell: bool, use_emergency: bool)`

#### **PRIORITY 0: Insecure Token First Sell** (Highest Priority)
- **Condition**: Token is in `INSECURE_TOKENS_FIRST_SELL_DETECTED` cache
- **Action**: **IMMEDIATE SELL** - Bypasses all other conditions
- **Reason**: Early detection of risky tokens
- **Logic**: 
  ```
  If insecure && first_sell_detected → SELL NOW
  ```

#### **PRIORITY 1: Low Interest Exit** (Market Activity)
- **Condition**: Low trading activity detected
- **Checks**:
  1. Net buy threshold: Per-second net(buy-sell) < threshold
  2. Volume threshold: Total volume < threshold SOL/s
- **Time Window**: Configurable lookback period (e.g., 60 seconds)
- **Logic**:
  ```
  If low_interest_lookback_secs > 0:
    If all last N seconds have net_buy < threshold → SELL
    OR if all last N seconds have volume < threshold → SELL
  ```

#### **PRIORITY 2: Max Hold Time**
- **Condition**: Token held longer than maximum time
- **Parameter**: `max_hold_time_secs` (configurable per token type)
- **Logic**:
  ```
  If time_held > max_hold_time → SELL
  ```

#### **PRIORITY 3: Stop Loss**
- **Condition**: PnL drops below stop loss threshold
- **Parameter**: Adaptive based on token type (see below)
- **Logic**:
  ```
  If pnl <= stop_loss → SELL
  ```
- **Example**: If stop_loss = -5% and current PnL = -6%, sell immediately

#### **PRIORITY 4: Trailing Stop**
- **Condition**: Price retracement exceeds trailing stop threshold
- **Activation**: Only activates when PnL >= activation percentage (e.g., 20%)
- **Dynamic Thresholds**: Based on current PnL level
  - >20% PnL: 5% trailing stop
  - >50% PnL: 10% trailing stop
  - >100% PnL: 30% trailing stop
  - >200% PnL: 100% trailing stop
- **Logic**:
  ```
  If pnl >= activation_percentage:
    trail_pct = get_trailing_stop_for_pnl(pnl)
    If retracement >= trail_pct → SELL
  ```
- **Example**: 
  - PnL = 30%, Highest price = $1.00, Current = $0.93
  - Retracement = (1.00 - 0.93) / 1.00 * 100 = 7%
  - Trail threshold for 30% PnL = 5%
  - Since 7% > 5% → SELL

#### **PRIORITY 5: Take Profit**
- **Condition**: PnL reaches take profit threshold
- **Parameter**: Adaptive based on token type
- **Logic**:
  ```
  If pnl >= take_profit → SELL
  ```
- **Example**: If take_profit = 15% and current PnL = 18%, sell

---

### Function: `get_adaptive_parameters_for_token()`

**Location**: `selling_strategy.rs`, line 939

Adaptive parameters adjust take profit, stop loss, and trailing stop based on token characteristics:

#### **Priority 1: Migration-Possible Tokens** (Marketcap > $60k)
- **Parameters** (env vars):
  - `MIGRATION_TAKE_PROFIT` (default: 12%)
  - `MIGRATION_STOP_LOSS` (default: -3%)
  - `MIGRATION_TRAILING_STOP_ACTIVATION` (default: 25%)
  - `MIGRATION_DYNAMIC_TRAILING_STOP_THRESHOLDS`
- **Rationale**: Tighter risk management before DEX migration

#### **Priority 2ایش Booming Tokens**
- **Parameters** (env vars):
  - `BOOMING_TOKEN_TAKE_PROFIT` (default: 15%)
  - `BOOMING_TOKEN_STOP_LOSS` (default: -5%)
  - `BOOMING_TOKEN_TRAILING_STOP_ACTIVATION` (default: 25%)
  - `BOOMING_TOKEN_DYNAMIC_TRAILING_STOP_THRESHOLDS`
- **Rationale**: Organic multi-trader activity, moderate volatility

#### **Priority 3: Pumping Tokens**
- **Parameters** (env vars):
  - `PUMPING_TAKE_PROFIT` (default: 15%)
  - `PUMPING_STOP_LOSS` (default: -5%)
  - `PUMPING_TRAILING_STOP_ACTIVATION` (default: 30%)
  - `PUMPING_DYNAMIC_TRAILING_STOP_THRESHOLDS`
- **Rationale**: Higher volatility, wider stops needed

#### **Default: Standard Tokens**
- Uses `SellingConfig` default values:
  - `take_profit` (from env)
  - `stop_loss` (from env)
  - `trailing_stop` config (from env)

---

### Internal Function: `execute_emergency_sell_internal()`

**Location**: `selling_strategy.rs`, line 1937

#### Step-by-Step Execution:

1. **Get Token Balance**:
   - Queries blockchain for current token balance
   - Exits if balance is zero

2. **Determine Protocol**:
   - Priority order:
     1. Use `parsed_data.dex_type` if available
     2. Use provided `protocol` parameter
     3. Fallback to stored metrics `protocol`
     4. Default to `PumpFun`

3. **Create Emergency Config**:
   - Slippage: 500% (whale emergency) or 300% (regular)
   - Amount: 100% of tokens
   - Direction: Sell

4. **Build Swap Instructions**:
   - **PumpFun**: Uses `pump.build_swap_from_parsed_data()`
   - **PumpSwap**: Uses `pump_swap.build_swap_from_parsed_data()`

5. **Get Recent Blockhash**:
   - Retrieves from `BlockhashProcessor`

6. **Execute Transaction**:
   - Uses `new_signed_and_send_zeroslot()` for maximum speed
   - Up to 3 internal retry attempts with 300ms delay

7. **Fallback to Jupiter API**:
   - If DEX selling fails, attempts Jupiter API fallback
   - Up to 3 retry attempts

8. **Verify and Cleanup**:
   - Waits 2 seconds after transaction
   - Verifies balance is zero
   - Cleans up tracking data:
     - Removes from `BOUGHT_TOKEN_LIST`
     - Removes from `TOKEN_METRICS`
     - Removes from `TOKEN_TRACKING`
     - Clears trade metrics

9. **Notification**:
   - Records sell for Telegram notification
   - Logs sell reason

---

### Function: `update_metrics()`

**Location**: `selling_strategy.rs`, line 761

Updates token metrics on every transaction:

- **Entry Price**: Set on first confirmed buy
- **Current Price**: Updated from transaction data
- **Highest Price**: Tracks peak price since entry
- **Lowest Price**: Tracks lowest price since entry
- **Price History**: Rolling window of last 20 prices
- **Liquidity**: Entry and current liquidity
- **Time Held**: Calculated from `buy_timestamp`
- **Protocol**: PumpFun or PumpSwap

---

### 1. Pumping Token First Sell
- **Location**: `sniper.rs`, line 4009
- **Trigger**: First sell detected for a pumping token
- **Action**: Immediate emergency sell
- **Priority**: Very high (similar to insecure token)

### 2. Target Wallet Copy Sell
- **Location**: `sniper.rs`, line 4087
- **Trigger**: Target wallet sells a token we own
- **Action**: Immediate copy sell (if enabled)
- **Status**: Currently disabled (DEX monitoring only)

### 3. Whale Emergency Sell
- **Parameter**: `is_whale_emergency: bool`
- **Difference**: 
  - Higher slippage tolerance (500% vs 300%)
  - Shorter timeout (30s vs 60s)
  - Faster execution priority

---

### TokenMetrics
```rust
pub struct TokenMetrics {
    pub entry_price: f64,
    pub highest_price: f64,
    pub lowest_price: f64,
    pub current_price: f64,
    pub volume_24h: f64,
    pub market_cap: f64,
    pub time_held: u64,
    pub last_update: Instant,
    pub buy_timestamp: u64,
    pub amount_held: f64,
    pub cost_basis: f64,
    pub price_history: VecDeque<f64>,
    pub volume_history: VecDeque<f64>,
    pub liquidity_at_entry: f64,
    pub liquidity_at_current: f64,
    pub protocol: SwapProtocol,
}
```

### Global State
- `TOKEN_METRICS`: DashMap<String, TokenMetrics> - All token metrics
- `TOKEN_TRACKING`: DashMap<String, TokenTrackingInfo> - Tracking info
- `SELL_REASONS`: DashMap<String, String> - Reasons for selling
- `SELL_LOCKS`: DashMap<String, ()> - Per-token sell locks

---

### SellingConfig
Loaded from environment variables via `SellingConfig::set_from_env()`

Key parameters:
- `take_profit`: Default take profit percentage
- `stop_loss`: Default stop loss percentage
- `trailing_stop`: Trailing stop configuration
- `low_interest_lookback_secs`: Time window for low interest checks
- `low_interest_net_buy_threshold_sol`: Net buy threshold
- `low_interest_volume_threshold_sol`: Volume threshold
- `time_based.max_hold_time_secs`: Maximum holding time

---

## Summary

**When Selling Starts**:
- Transaction stream monitoring (primary)
- Periodic token monitoring
- Manual/external triggers

**Decision Function**: `evaluate_sell_conditions()` 
- Checks conditions in priority order
- Returns (should_sell, use_emergency)

**Execution Function**: `unified_emergency_sell()`
- Handles race conditions
- Manages retries and timeouts
- Executes swap via DEX or Jupiter API
- Verifies and cleans up

**Key Selling Conditions** (in priority order):
1. Insecure token first sell (highest priority)
2. Low interest exit
3. Max hold time exceeded
4. Stop loss triggered
5. Trailing stop triggered
6. Take profit reached
